src/main.rs

```diff
@@ -63,6 +75,213 @@ async fn init_maker(
     }
 }
 
+fn init_rate_updates(
+    update_interval: Duration,
+) -> (
+    impl Future<Output = comit::Never> + Send,
+    Receiver<anyhow::Result<MidMarketRate>>,
+) {
+    let (mut sender, receiver) = futures::channel::mpsc::channel::<anyhow::Result<MidMarketRate>>(
+        ENSURED_CONSUME_ZERO_BUFFER,
+    );
+
+    let future = async move {
+        loop {
+            let rate = get_btc_dai_mid_market_rate().await;
+
+            if let Err(e) = sender.try_send(rate) {
+                tracing::trace!("Error when sending rate update from sender to receiver.")
+            }
+
+            Delay::new(update_interval).await;
+        }
+    };
+
+    (future, receiver)
+}
+
+fn init_bitcoin_balance_updates(
+    update_interval: Duration,
+    wallet: bitcoin_wallet::Wallet,
+) -> (
+    impl Future<Output = comit::Never> + Send,
+    Receiver<anyhow::Result<bitcoin::Amount>>,
+) {
+    let (mut sender, receiver) = futures::channel::mpsc::channel::<anyhow::Result<bitcoin::Amount>>(
+        ENSURED_CONSUME_ZERO_BUFFER,
+    );
+
+    let future = async move {
+        loop {
+            let balance = wallet.balance().await;
+
+            if let Err(e) = sender.try_send(balance) {
+                tracing::trace!("Error when sending rate update from sender to receiver.")
+            }
+
+            Delay::new(update_interval).await;
+        }
+    };
+
+    (future, receiver)
+}
+
+fn init_dai_balance_updates(
+    update_interval: Duration,
+    wallet: ethereum_wallet::Wallet,
+) -> (
+    impl Future<Output = comit::Never> + Send,
+    Receiver<anyhow::Result<dai::Amount>>,
+) {
+    let (mut sender, receiver) =
+        futures::channel::mpsc::channel::<anyhow::Result<dai::Amount>>(ENSURED_CONSUME_ZERO_BUFFER);
+
+    let future = async move {
+        loop {
+            let balance = wallet.dai_balance().await;
+
+            if let Err(e) = sender.try_send(balance.map(|balance| balance.into())) {
+                tracing::trace!("Error when sending rate update from sender to receiver.")
+            }
+
+            Delay::new(update_interval).await;
+        }
+    };
+
+    (future, receiver)
+}
+
+async fn execute_swap(sender: Sender<FinishedSwap>) -> anyhow::Result<()> {
+    let position: Position =
+        todo!("decision what kind of what swap it is hbit->herc20 or herc20->hbit");
+
+    let taker: Taker = todo!("Taker has to be available after execution, e.g. load from db");
+
+    match position {
+        Position::Sell => {
+            let beta_params: hbit::Params = unimplemented!();
```

Comment:
This unimplemented and probably the `todo!`s above are the reason for the ownership not being enforced :)

Reply:
Here is a playground of a smaller example where the ownership is enforced:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d1c2d079d84518e90121f6a3023412b1